### PR TITLE
INS-420

### PR DESCRIPTION
**Pull request by Fermionic-Lyu**
## Summary

Fixes [INS-420](https://linear.app/insforge/issue/INS-420/triage-why-cloud-projects-are-saving-insforge-logs-too): cloud projects were writing `/insforge-logs/insforge.logs.jsonl` inside the container even though logs already ship to CloudWatch via the awslogs driver and are read back through CloudWatchProvider. The file was an unread, unrotated duplicate growing the container filesystem.

Root cause: the winston logger had an unconditional File transport, and the Dockerfile sets `LOGS_DIR=/insforge-logs` for every containerized deployment.

While triaging, this also surfaced that self-hosted dashboard logs have been dead since the Vector sidecar was removed in April — `LocalFileProvider` skipped every line lacking the Vector `appname` field, which winston never writes.

## Changes

- `backend/src/utils/logger.ts`: only add the File transport when not in a cloud environment; create the logs directory eagerly; rotate the file (20MB, 2 files, tailable) so self-hosted disks cannot grow unbounded either.
- `backend/src/providers/logs/local.provider.ts`: parse winston's native line shape — app logs as `level - message`, HTTP request logs as nginx-style lines (matching the old Vector output), error logs with `Error:` / `Stack Trace:` appended like the CloudWatch display. Legacy Vector lines still parse. The provider now reads from `appConfig.server.logsDir`, the same directory winston writes to (the two sides previously had different bare-metal defaults).

No insforge-standalone changes needed — cloud gating keys off `AWS_INSTANCE_PROFILE_NAME`, which provisioning already sets.

## Testing

- New `local-log-provider.test.ts`: winston and legacy Vector line parsing, request/error formatting, unknown-line skipping, timestamp filter, search.
- New `logger-transports.test.ts`: file transport present when self-hosted and absent in cloud, plus a real round-trip (winston write, LocalFileProvider read) so writer and reader formats cannot drift apart.
- Typecheck, eslint, prettier clean; remaining full-suite failures are pre-existing storage/S3 tests (missing `lru-cache` types), verified to fail identically on a clean tree.

Known limit: self-host regains backend logs; postgres/postgrest/function sources still have no writer without a log shipper (separate follow-up).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Stops cloud containers from writing local log files and restores self‑hosted dashboard logs. Fixes INS‑420 by removing duplicate on‑disk logs in cloud and aligning the reader/writer so logs show up again.

- **Bug Fixes**
  - Gate the `winston` file transport to non‑cloud; eagerly create the logs dir; rotate (20MB, 2 files, tailable); flatten `Error` metadata before JSONL; warn if file logging init fails.
  - `LocalFileProvider`: use `appConfig.server.logsDir`; parse both `winston` JSONL and legacy Vector; format HTTP lines only when `method` and `duration` exist; include error/stack; read legacy severity from `metadata.level`.
  - Cloud behavior: no local log file; logs stay in CloudWatch to avoid container disk growth. Self‑hosted instances get rotated on‑disk logs.
  - Tests: transport gating, rotation config, real `Error` round‑trip, legacy/winston parsing; storage URL versioning suite mocks `isCloudEnvironment` to report cloud and skip file logging.

<sup>Written for commit 36aa1032620c1c37590b12ca487ffbaf48b64177. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1621?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

> [!NOTE]
> ### Add winston file transport logging and update `LocalFileProvider` to parse winston log entries
> - Adds a rotating JSONL file transport (`insforge.logs.jsonl`, max 20MB, 2 files) to the winston logger in non-cloud environments via [`logger.ts`](https://github.com/InsForge/InsForge/pull/1621/files#diff-a1b886b6dd6d43dd9e59b8abb92e3fc8f023eaff455660b387aec996fe09cbb9); cloud environments continue to log to console only.
> - Adds a `formatEventMessage` helper to [`LocalFileProvider`](https://github.com/InsForge/InsForge/pull/1621/files#diff-30a179baec282c28f60f8c996a63cba329f6e18f077de09f140a34d4ce0a723b) that parses both legacy Vector-shaped entries and new winston file transport entries, producing nginx-style lines for HTTP requests and appending error/stack for error-level logs.
> - Updates `LocalFileProvider.initialize()` to use `appConfig.server.logsDir` instead of `process.env.LOGS_DIR`.
> - Adds unit tests for both the logger transport selection and the updated `LocalFileProvider` parsing logic.
> - Behavioral Change: `readLogsFromFile` now includes winston-formatted lines in results; previously only Vector-shaped lines (with `appname`) were returned.
>
> 
> #### Changes since #1621 opened
>
> - Extended mock for `@/utils/environment.js` module to return `false` for `isCloudEnvironment()` method alongside existing `getApiBaseUrl` mock [bf91ac0]
> - Implemented error flattening in the logging pipeline [55b0eb1]
> - Modified `LocalFileProvider.formatEventMessage` to derive severity from `metadata.level` for legacy Vector-shaped entries and tightened Winston HTTP request detection [55b0eb1]
> - Updated file transport configuration error handling and added rotation settings validation in tests [55b0eb1]
> - Added test coverage for error flattening and format detection improvements [55b0eb1]
> - Changed mock return value of `isCloudEnvironment` from `false` to `true` in `vi.mock('@/utils/environment.js')` block [36aa103]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized dc96bc1.</sup>
> 
>
> 

## Summary by CodeRabbit

* **New Features**
  * Log viewing now supports both legacy and current log formats, so more application logs can be read successfully.
  * Logging now works more reliably across environments, including cloud deployments.

* **Bug Fixes**
  * Improved log parsing to skip only unsupported or invalid lines while keeping valid entries.
  * Fixed log storage path handling so local logs are saved from the configured app settings.

## E2E

Deterministic Fixture E2E passed against test image `v2.2.5-ins-420`: [run 28553763736](https://github.com/InsForge/agent-e2e/actions/runs/28553763736) (passed on all three revisions of this branch) ([image build](https://github.com/InsForge/InsForge/actions/runs/28551511433)). No `agent-e2e` fixture changes needed — no fixture asserts the logs API, and the cloud log-read path (CloudWatch) is unchanged.

### CODE REVIEW

**Review by jwfing — COMMENTED** [review]
## Summary

Cleanly stops cloud containers from writing an unread on-disk log file and restores self-hosted dashboard logs by aligning the winston writer format with the `LocalFileProvider` reader; well-scoped and well-tested, no blocking issues.

## Requirements context

No matching spec/plan under `docs/superpowers/` — that directory only contains an unrelated e2e-testing skill design (`specs/2026-06-29-e2e-testing-skill-design.md`, `plans/2026-06-29-e2e-testing-skill.md`). Assessed against the PR description and linked issue INS-420 alone. The stated intent (cloud stops writing `/insforge-logs/insforge.logs.jsonl`; self-hosted dashboard logs parse again; disks bounded via rotation) is coherent and matches the diff.

## Findings

### Critical
(none)

### Suggestion

**Functionality — request-log detection is a field-name heuristic** (`backend/src/providers/logs/local.provider.ts:102`)
Winston lines are classified as HTTP requests solely by `metadata.duration !== undefined`. The request logger (`backend/src/server.ts:143-152`) does emit exactly `method/path/status/size/duration/ip/userAgent`, so this works today. But any future `logger.info('...', { duration })` — a plausible field name for timing an operation — would be silently reformatted into a garbled nginx-style line with `undefined` in most slots. Consider gating on `message === 'HTTP Request'` (or presence of `method` + `path`) so the discriminator is intent, not an incidental field name.

**Software engineering — error serialization is only tested for the string case** (`backend/src/providers/logs/local.provider.ts:116-122`, test `backend/tests/unit/local-log-provider.test.ts:33-38`)
The parser appends `Error:`/`Stack Trace:` from `metadata.error`/`metadata.stack`, and the test exercises `{ error: 'boom', stack: '...' }`. But the prevailing codebase pattern is `logger.error('msg', { error })` where `error` is a raw `Error` object (e.g. `usage.service.ts:59`, `secret.service.ts:80`, and many more), not `String(error)`. `winston.format.errors()` only lifts an Error at the top level of `info`, not one nested under `metadata.error`, so on disk `metadata.error` serializes via `JSON.stringify` to `{}` and the dashboard renders `Error: [object Object]`/`Error: {}` with no message or stack. Since the whole point of this PR is legible dashboard logs, it's worth either normalizing the error field on write or adding a test with a real `Error` object to make the actual display behavior explicit.

### Information

**Security — secrets still reach the on-disk file on self-hosted (pre-existing, out of scope).** `seed.ts` logs `API key generated: …` / `Anon key generated: …` at `info` level, so on self-hosted these now land in the rotated `insforge.logs.jsonl`. This is unchanged by the PR (the File transport always existed) and cloud behavior actually improves (no file at all). Flagging only as a heads-up; a separate PR could drop those to `debug` or redact.

**Performance — `formatEventMessage` now runs before the timestamp filter** (`local.provider.ts:156-163`). Message formatting moved ahead of the `logTime < beforeMs` check, so it's computed for lines that are then discarded. Negligible at these volumes; noting for completeness.

**Style — silent `catch {}` on mkdir** (`backend/src/utils/logger.ts:38-40`). Swallowing the directory-creation failure is intentional and commented (console logging still works); reasonable, no change requested.

**Reader/writer alignment looks correct.** `tailable: true` keeps the newest entries in the base `insforge.logs.jsonl` that `LocalFileProvider` reads (`logger.ts:25`), the logger-level `winston.format.timestamp()` populates `info.timestamp` before the transport `printf` consumes it, and both sides now resolve `appConfig.server.logsDir` (`local.provider.ts:20`), fixing the prior bare-metal default mismatch. Cloud gating via `isCloudEnvironment()` is evaluated at module load, which is correct since `AWS_INSTANCE_PROFILE_NAME` is present at container start.

## Verdict

**approved** (informational — human approval via the separate approve flow). Zero Critical findings: the fix is correct, scoped to the logging path, and backed by focused unit tests plus a round-trip test that guards writer/reader drift. The two Suggestions (request-log discriminator robustness, real-`Error` serialization test) are worth addressing but non-blocking.

File: backend/src/providers/logs/local.provider.ts [hunk 1]
```diff
@@ -71,6 +73,59 @@ export class LocalFileProvider extends BaseLogProvider {
...
+      // For error logs, include error and stack in eventMessage to match CloudWatch display
+      let eventMessage = String(log.event_message ?? '');
+      if (log.level === 'error' && log.error) {
+        eventMessage = `${eventMessage}\n\nError: ${log.error}`;
+        if (log.stack) {
+          eventMessage += `\n\nStack Trace:\n${log.stack}`;
+        }
+      }
+      return eventMessage;
+    }
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/logs/local.provider.ts:96]: Fixed in 55b0eb1cc — the Vector branch now reads metadata.level (with a top-level fallback), and there's a fixture covering a legacy error line.

File: backend/src/providers/logs/local.provider.ts [hunk 2]
```diff
@@ -71,6 +73,59 @@ export class LocalFileProvider extends BaseLogProvider {
...
+
+    // Winston file transport entries: { id, timestamp, message, level, metadata }
+    if (log.message !== undefined && log.level && log.timestamp) {
+      const metadata = (log.metadata ?? {}) as Record<string, unknown>;
+
+      // HTTP request logs (see the request logger in server.ts) — format as an
+      // nginx-style line, matching what the Vector pipeline used to produce
+      if (metadata.duration !== undefined) {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/logs/local.provider.ts]: Fixed in 55b0eb1cc — the guard now requires metadata.method alongside duration, with a test asserting a timing-only log stays in app-log format.

File: backend/src/utils/logger.ts [hunk 3]
```diff
@@ -1,9 +1,45 @@
...
+            message,
+            level,
+            metadata,
+          });
+        }),
+      })
+    );
+  } catch {
+    // Logs directory is not writable — console logging still works.
+  }
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/utils/logger.ts]: Fixed in 55b0eb1cc — the catch now emits a console.warn with the directory and error.

[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/utils/logger.ts]: Fixed in 55b0eb1cc — init failures now warn on the console instead of being swallowed.

File: backend/src/providers/logs/local.provider.ts [hunk 4]
```diff
@@ -71,6 +73,59 @@ export class LocalFileProvider extends BaseLogProvider {
...
+   * is not a log entry we understand. Handles both line shapes found in these
+   * files: entries written by the winston file transport (utils/logger.ts) and
+   * legacy entries shipped by the Vector sidecar before it was removed.
+   */
+  private formatEventMessage(log: Record<string, unknown>): string | null {
+    // Legacy Vector-transformed entries carry an appname field
+    if (log.appname) {
+      // For error logs, include error and stack in eventMessage to match CloudWatch display
+      let eventMessage = String(log.event_message ?? '');
+      if (log.level === 'error' && log.error) {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/logs/local.provider.ts]: Fixed in 55b0eb1cc — reads metadata.level with a fallback to log.level, plus a fixture for a legacy Vector error line.

File: backend/tests/unit/logger-transports.test.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      process.env.AWS_INSTANCE_PROFILE_NAME = originalProfile;
+    } else {
+      delete process.env.AWS_INSTANCE_PROFILE_NAME;
+    }
+  });
+
+  it('writes insforge.logs.jsonl when self-hosted', async () => {
+    const logger = await importLogger();
+
+    expect(logger.transports.some((t) => t instanceof winston.transports.File)).toBe(true);
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/tests/unit/logger-transports.test.ts]: Done in 55b0eb1cc — the self-hosted transport test now asserts maxsize, maxFiles, and tailable.

File: backend/tests/unit/logger-transports.test.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    for (let i = 0; i < 20 && matches.length === 0; i++) {
+      const { logs } = await provider.getLogsBySource('insforge.logs');
+      matches = logs.filter((l) => l.eventMessage === 'info - Round trip works');
+      if (matches.length === 0) {
+        await new Promise((r) => setTimeout(r, 50));
+      }
+    }
+
+    expect(matches).toHaveLength(1);
+  });
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/tests/unit/logger-transports.test.ts:87]: Keeping the poll: awaiting 'finish' via logger.end() breaks here because LocalFileProvider.initialize logs through the same logger instance ('write after end'). Flushes are ms-scale; the 1s budget has plenty of headroom.

**Comment by Fermionic-Lyu** [comment]
Addressed the review findings in 55b0eb1cc: raw Error objects in metadata are now flattened to {message, stack} before hitting disk (with a real-Error round-trip test), the HTTP-request discriminator also requires metadata.method, legacy Vector severity is read from metadata.level, mkdir failures warn instead of vanishing, and rotation config is asserted in tests. Left the poll-based flush wait as-is — logger.end() breaks because LocalFileProvider.initialize logs through the same logger.

**Review by jwfing — COMMENTED** [review]
## Review: INS-420 — gate winston file transport to non-cloud + restore self-hosted dashboard logs

**Summary:** Cleanly stops cloud containers from writing an unread, unrotated `insforge.logs.jsonl`, and repairs self-hosted log reading by teaching `LocalFileProvider` winston's native line shape (while still parsing legacy Vector lines) — well-scoped and well-tested, no blocking issues.

### Requirements context
No spec/plan under `docs/superpowers/{specs,plans}/` matches this PR — the only docs there cover the e2e-testing skill. Assessed against the PR description and linked issue INS-420 alone. Intent (cloud stops writing a duplicate on-disk log; self-hosted regains dashboard logs by aligning reader/writer formats) is clear and internally consistent.

### Findings

#### Critical
(none)

#### Suggestion
- **Software engineering / test hygiene — `backend/tests/unit/storage-url-versioning.test.ts:5-9`.** The environment mock forces `isCloudEnvironment: () => false`, but this test does not mock `@/infra/config/app.config`. So when `logger.ts` is pulled in transitively it runs the new module-load-time `fs.mkdirSync(appConfig.server.logsDir)` (real default `process.cwd()/logs`, see `app.config.ts:135`) and opens a File transport that writes `logs/insforge.logs.jsonl` into the working tree during an unrelated storage test. Mocking it to `() => true` (this test doesn't exercise file logging) avoids the stray directory/file side-effect; alternatively mock `app.config` to a temp `logsDir`. Non-blocking.

#### Information
- **Functionality / format parity — `backend/src/providers/logs/local.provider.ts:106-116` vs `backend/src/providers/logs/cloudwatch.provider.ts:595-606`.** The nginx-style request line here joins raw `metadata` values directly, while the CloudWatch provider wraps `status`/`size`/`duration` in `fmt()` (undefined/null → `''`). `Array.prototype.join` already coerces `undefined`/`null` to `''`, so output is identical for the fields the request logger writes (`server.ts:143-152` always populates all of them) — noting only for reviewer confidence that cloud and self-hosted displays stay byte-identical.
- **Security — no security-relevant regressions.** Stack traces are gated on `NODE_ENV !== 'production'` in *both* `winston.format.errors` and the new `flattenErrors` replacer (`logger.ts:14-20, 65`), so path/module-structure leakage is consistently suppressed in prod. No new user input reaches SQL/shell/HTTP; the `getValidatedPath` traversal guard (`local.provider.ts:40-50`) is untouched; no secrets/PII newly logged or returned.
- **Performance — net improvement.** Cloud no longer writes the duplicate file at all; self-hosted growth is now bounded by rotation (20MB × 2, `tailable`), and `tailable` correctly keeps the newest entries in the base file `LocalFileProvider` reads. The full-file scan in `readLogsFromFile`/`searchLogs` is pre-existing and now operates on a size-bounded file — no new N+1 or hot-path work introduced.

### Notes on strengths
- The `logger-transports.test.ts` round-trip (winston writes → `LocalFileProvider` reads, including a real `new Error()`) directly guards against writer/reader format drift — exactly the class of bug this PR is fixing. Edge cases (timing metadata not mistaken for an HTTP line, unknown-shape/invalid-JSON skipping, legacy `metadata.level` severity, `beforeTimestamp` filter) are all covered.
- Scope is tight: 3 source hunks + 2 test files, no drive-by refactors. Cloud gating reuses the existing `AWS_INSTANCE_PROFILE_NAME` signal, so no insforge-standalone/provisioning change is required.

### Verdict
**approved** (informational — zero Critical findings). The single Suggestion is optional test hygiene. Posting as a COMMENT; explicit GitHub green-check approval remains a separate human action.

**Comment by Fermionic-Lyu** [comment]
Confirmed the test-hygiene finding — running that suite really did leave a 111KB logs/insforge.logs.jsonl in the working tree (masked by the backend/logs/* gitignore entry). Fixed in 36aa10326: the mock now reports cloud so the file transport is skipped, and I verified the suite passes without creating the directory.

**Review by jwfing — APPROVED** [review]
LGTM - approved.
